### DISCUSSION TITLE
Supporting @section, somehow

### DISCUSSION THREAD

**Discussion by kftrans** [adoconnection/RazorEngineCore#61] [Q&A]
I'm porting an app that uses the older RazorEngine library. RazorEngineCore is excellent, particularly the cleaner approach for building custom templates that support @Html etc. The example at https://github.com/adoconnection/RazorEngineCore/wiki/@Include-and-@Layout provides good information for other kinds of support. However, I'm struggling to see any reasonably clean/robust way to support @section. Has anyone tried this?
Thanks.

**Comment by adoconnection** [comment]
we should go with ```directives```

there is a native ```SectionDirective.Register(builder);``` in RazorEngine
https://github.com/aspnet/Razor/blob/5439cfe540084edd673b7ed626f2ec9cf3f13b18/src/Microsoft.AspNetCore.Razor.Language/Extensions/SectionDirective.cs

```cs CreateAndCompileToStream
            RazorProjectEngine engine = RazorProjectEngine.Create(
                RazorConfiguration.Default,
                RazorProjectFileSystem.Create(@"."),
                (builder) =>
                {
                    builder.SetNamespace(options.TemplateNamespace);
                    SectionDirective.Register(builder);
                });
```

I think we can go and implement directives in next releases so the engine will support what it support right now and can be extended with custom keywords

**Reply by kftrans** [reply]
Thanks for that. I've tried to create the most minimal set of changes to allow @section and @RenderSection to be supported, using the existing/native classes. I've put those in a PR, with an example. Not sure if it's the kind of thing you had in mind and/or if there's a better way, but any feedback appreciated.

**Comment by adoconnection** [comment]
Hi, thanks for PR!
I was thinking about directives in general, as razor let us to create any directives we want.
So to avoid future breaking changes we have to think at least about sections, model and inject.

The list of things to consider:
* sections and layouts (your PR)
* inject and ability to use any given IoC
* easy localization \ resources
* allow @model keyword  instead of @inherits

I like the idea to use readymade Razor directives, 
I would like to avoid referencing those directives directly to keep namespace clean, i.e. to have RazorEngineCore reference only in calling code.

**Reply by wdcossey** [reply]
```csharp
        //Taken from Microsoft.AspNetCore.Razor.Language.Extensions
        public static class SectionDirective
        {
            public static readonly DirectiveDescriptor Directive = DirectiveDescriptor.CreateDirective(
                "section",
                DirectiveKind.RazorBlock,
                builder =>
                {
                    builder.AddMemberToken("SectionName", "The name of the section.");
                    builder.Description = "Define a section to be rendered in the configured layout page.";
                });

            public static void Register(RazorProjectEngineBuilder builder)
            {
                if (builder == null)
                {
                    throw new ArgumentNullException(nameof(builder));
                }

                builder.AddDirective(Directive, FileKinds.Legacy, FileKinds.Component);
                builder.Features.Add(new SectionDirectivePass()); // You can get this from Microsoft.AspNetCore.Razor.Language.Extensions
                builder.AddTargetExtension(new SectionTargetExtension()); // You can get this from Microsoft.AspNetCore.Razor.Language.Extensions
            }
        }
```

```csharp
        protected virtual RazorProjectEngine CreateEngine(RazorEngineCompilationOptions options)
        {
            return RazorProjectEngine.Create(
                RazorConfiguration.Default,
                RazorProjectFileSystem.Create(@"."),
                (builder) =>
                {
                    builder.SetNamespace(options.TemplateNamespace);
                    SectionDirective.Register(builder);
                });
        }
```

**Comment by adr-v** [comment]
Hi guys, is there any update on this?
Are sections supported or is there any plan to add support?
